# Mitogenomes
Tools, scripts, and commands to conduct bioinformatic analyses on mitochondrial genomes.

To install packages and softwares, see [this page](https://github.com/depaulats/Mitogenomes/blob/main/intalling_packages_conda.md).

To dump, clean, map and assemble reads, see [this page](https://github.com/depaulats/Mitogenomes/blob/main/dump_clean_map_assemble.md). If you prefer mapping reads using Bowtie2

To extract data fom Genbank files, use the python scripts:
- [gbex_translation.py](https://github.com/depaulats/Mitogenomes/blob/main/gbex_tranlation.py) → for amino acid data from translated proteins
- [gbex_rrnaseq.py](https://github.com/depaulats/Mitogenomes/blob/main/gbex_rrnaseq.py) → for nucleotide data from rRNAs
- [gbex_gene.py](https://github.com/depaulats/Mitogenomes/blob/main/gbex_gene.py) → for nucleotide data from all genes (including tRNAs) 

To conduct a phylogenomic analysis, see [this page](https://github.com/depaulats/Mitogenomes/blob/main/phylogenomics.md).

